finat/argyris.py

```diff
@@ -48,47 +89,53 @@ def basis_transformation(self, coordinate_mapping):
-            V[18+e, 6*v1id] = 15/8 * (foo / pel[e])
-
-            # vertex derivatives
-            for i in (0, 1):
-                V[18+e, 6*v0id+1+i] = -7/16*foo*pts[e, i]
-                V[18+e, 6*v1id+1+i] = V[18+e, 6*v0id+1+i]
-
-            # second derivatives
-            tau = [pts[e, 0]*pts[e, 0],
-                   2*pts[e, 0]*pts[e, 1],
-                   pts[e, 1]*pts[e, 1]]
-
-            for i in (0, 1, 2):
-                V[18+e, 6*v0id+3+i] = -1/32 * (pel[e]*foo*tau[i])
-                V[18+e, 6*v1id+3+i] = 1/32 * (pel[e]*foo*tau[i])
-
-            V[18+e, 18+e] = (rns[e, 0]*(J[0, 0]*pns[e, 0] + J[1, 0]*pns[e, 1])
-                             + rns[e, 1]*(J[0, 1]*pns[e, 0] + J[1, 1]*pns[e, 1]))
+        q = self.degree - 4
+        if self.variant == "integral":
+            _edge_transform(V, voffset, self.cell, q, coordinate_mapping, avg=self.avg)
+        else:
+            rns = coordinate_mapping.reference_normals()
+            pns = coordinate_mapping.physical_normals()
+            pts = coordinate_mapping.physical_tangents()
+            pel = coordinate_mapping.physical_edge_lengths()
+            for e in sorted(top[1]):
+                nhat = partial_indexed(rns, (e, ))
+                n = partial_indexed(pns, (e, ))
+                t = partial_indexed(pts, (e, ))
+                Bn = J @ nhat
+                Bnt = Bn @ t
+                Bnn = Bn @ n
+
+                s = len(top[0]) * voffset + e
+                v0id, v1id = (v * voffset for v in top[1][e])
+                V[s, s] = Bnn
+
+                # vertex points
+                V[s, v0id] = -15/8 * Bnt / pel[e]
+                V[s, v1id] = 15/8 * Bnt / pel[e]
+
+                # vertex derivatives
+                for i in range(sd):
+                    V[s, v0id+1+i] = -7/16 * Bnt * t[i]
+                    V[s, v1id+1+i] = V[s, v0id+1+i]
+
+                # second derivatives
+                tau = [t[0]*t[0], 2*t[0]*t[1], t[1]*t[1]]
+                for i in range(len(tau)):
+                    V[s, v0id+3+i] = -1/32 * (pel[e] * Bnt * tau[i])
+                    V[s, v1id+3+i] = 1/32 * (pel[e] * Bnt * tau[i])
 
         # Patch up conditioning
         h = coordinate_mapping.cell_size()
-
-        for v in range(3):
-            for k in range(2):
-                for i in range(21):
-                    V[i, 6*v+1+k] = V[i, 6*v+1+k] / h[v]
-            for k in range(3):
-                for i in range(21):
-                    V[i, 6*v+3+k] = V[i, 6*v+3+k] / (h[v]*h[v])
-        for e in range(3):
-            v0id, v1id = [i for i in range(3) if i != e]
-            for i in range(21):
-                V[i, 18+e] = 2*V[i, 18+e] / (h[v0id] + h[v1id])
+        for v in sorted(top[0]):
+            for k in range(sd):
+                V[:, voffset*v+1+k] *= 1 / h[v]
+            for k in range((sd+1)*sd//2):
+                V[:, voffset*v+3+k] *= 1 / (h[v]*h[v])
+
+        eoffset = 2 * q - 1
```

Comment:
The exact nature of the edge dof scaling here is going to depend on the variant of integral vs point values.  We should get that straightened out (tests of mass matrix conditioning will reveal this).